service/src/BatchServer.hpp

```diff
@@ -0,0 +1,167 @@
+            ///        created.
+            virtual int server_pid(void) const = 0;
+            /// @return The key used to identify the server
+            ///        connection: a substring in interprocess shared
+            ///        memory keys used for communication.
+            virtual std::string server_key(void) const = 0;
+            /// @brief Supports the D-Bus interface for stopping a
+            ///        batch server.
+            ///
+            /// This function is called directly by geopmd in order to
+            /// end a batch session and kill the batch server process
+            /// created by start_batch_server().
+            virtual void stop_batch(void) = 0;
+            /// @brief Returns true if the batch server is running
+            virtual bool is_active(void) const = 0;
+    };
+
+    class BatchServerImp : public BatchServer
+    {
+        public:
+            BatchServerImp(int client_pid,
+                           const std::vector<geopm_request_s> &signal_config,
+                           const std::vector<geopm_request_s> &control_config);
+            BatchServerImp(int client_pid,
+                           const std::vector<geopm_request_s> &signal_config,
+                           const std::vector<geopm_request_s> &control_config,
+                           PlatformIO &pio,
+                           std::shared_ptr<BatchStatus> batch_status,
+                           std::shared_ptr<SharedMemory> signal_shmem,
+                           std::shared_ptr<SharedMemory> control_shmem);
+            virtual ~BatchServerImp();
+            int server_pid(void) const override;
+            std::string server_key(void) const override;
+            void stop_batch(void) override;
+            bool is_active(void) const override;
+            void run_batch(void);
+            /// @brief Fork a process that runs two functions and
+            ///        block until the first function completes.
+            int fork_with_setup(std::function<void(void)> setup,
+                                std::function<void(void)> run);
+        private:
+            void push_requests(void);
+            void read_and_update(void);
+            void update_and_write(void);
+            void create_shmem(void);
+            void check_invalid_signal(void);
+            void check_return(int ret, const std::string &func_name) const;
+
+            int m_client_pid;
+            std::vector<geopm_request_s> m_signal_config;
+            std::vector<geopm_request_s> m_control_config;
+            PlatformIO &m_pio;
+            std::shared_ptr<SharedMemory> m_signal_shmem;
+            std::shared_ptr<SharedMemory> m_control_shmem;
+            std::shared_ptr<BatchStatus> m_batch_status;
+            std::string m_server_key;
+            int m_server_pid;
+            bool m_is_active;
+            std::vector<int> m_signal_idx;
+            std::vector<int> m_control_idx;
```

Comment:
Would be good to include doxygen comments to describe what exactly is stored in here.